### PR TITLE
Fix off-by-one in MoonshineStreaming sliding_window_mask_function right window

### LINKED ISSUE #46305: MoonshineStreaming: sliding_window_mask_function off-by-one — right_window_size=4 attends to 3 future frames instead of 4
### System Info

- transformers version: 5.9.0
- Platform: Linux-6.8.0-100-generic-x86_64-with-glibc2.39
- Python version: 3.12.3
- PyTorch version: 2.12.0

### Who can help?

@eustlb 

### Information

`sliding_window_mask_function` in `modeling_moonshine_streaming.py` (line 349) uses strict less-than (`< right_window_size`) for the right context, which means `right_window_size=4` only attends to 3 future frames instead of 4.

The Moonshine v2 paper (arXiv:2602.12241) explicitly states:

`"a right window of w_right=4 implies an algorithmic lookahead of 4×20 ms = 80 ms"`

At 50 Hz, 4 future frames = 80 ms. But the current code gives 3 future frames = 60 ms.

- [ ] The official example scripts
- [x] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [x] My own task or dataset (give details below)

### Reproduction

from transformers.models.moonshine_streaming.modeling_moonshine_streaming import sliding_window_mask_function

fn = sliding_window_mask_function((16, 4))

q = 10
future = [kv for kv in range(q + 1, 20) if fn(0, 0, q, kv)]
print(f"Future frames attended: {future}")
Actual:   [11, 12, 13]     — 3 frames, 60ms
Expected: [11, 12, 13, 14] — 4 frames, 80ms (per paper)

### Expected behavior

With sliding_window_mask_function((16, 4)), a query at position 10 should attend to future positions {11, 12, 13, 14} — 4 future frames, matching the paper's definition of w_right=4

### PR DESCRIPTION
**Pull request by namantiwari2002**
[CI](https://transformers-ci.lor-e.huggingface.cool/d/pytest-observability-by-pr/pytest-observability-branch?var-pr=47162)

## What

`sliding_window_mask_function` in `modeling_moonshine_streaming.py` (and its
`modular_` source) builds the right (future) side of the sliding-window
attention mask with a strict inequality:

```python
right_mask = (dist < 0) & (-dist < right_window_size)
```

With `dist = q_idx - kv_idx`, a future key is `dist < 0`, and `-dist` is the
number of frames the key is ahead of the query. The strict `<` drops the
frame exactly `right_window_size` steps ahead, so a configured right window of
`w_right = 4` only attends to 3 future frames instead of 4.

Concretely, `sliding_window_mask_function((16, 4))` for a query at position 10
returned future positions `{11, 12, 13}` instead of the expected
`{11, 12, 13, 14}`.

This does not match the Moonshine v2 definition, where `w_right = 4` implies an
algorithmic lookahead of `4 x 20 ms = 80 ms` (4 frames at 50 Hz), not 60 ms.

## Change

Use `<=` for the right-window comparison so the full right window is included:

```python
right_mask = (dist < 0) & (-dist <= right_window_size)
```

The left window and causal semantics are unchanged. The same generated copy in
`gemma4/modeling_gemma4.py` (which reuses this function via `modular`) is
updated to stay in sync.

Fixes #46305

## How tested

Added a unit test (`MoonshineStreamingSlidingWindowMaskTest`) asserting that a
query at position 10 with window `(16, 4)` attends to exactly
`{11, 12, 13, 14}`, that the frame 5 steps ahead is excluded, and that the left
window is unaffected.

```
python -m pytest tests/models/moonshine_streaming/test_modeling_moonshine_streaming.py::MoonshineStreamingSlidingWindowMaskTest -v
```

The fix logic was confirmed with a standalone reproduction script mirroring
the issue's repro case — output changed from `[11, 12, 13]` to `[11, 12, 13, 14]`.

### CODE REVIEW

**Comment by Rocketknight1** [comment]
cc @eustlb @ebezzam since this is an audio sliding window issue. I think the bug is legit, and this code is imported into Gemma4! It's only used for Gemma4 audio, though, and I think the actual Gemma4 configs might mean it doesn't have much effect, but it's still probably important enough to check/fix.
